Sort imports, add check in CI

pawamoy opened the pull request:
For issue #1625

First commit adds the configuration for isort in setup.cfg, the tox env in tox.ini, and the job in travis config.

Next commit will actually fix the imports, to see if CI passes.

pawamoy commented:
I think #1618 should be done first, to avoid ordering imports now and re-ordering them after #1628.

pawamoy commented:
Job correctly failing, pushing the fix :slightly_smiling_face:

ramnes commented:
Yisss! I don't like the 120 line length though, but I guess we'll discuss this in the issue about code formatting.

pawamoy commented:
I can change it, no problem. Thing is, flake8 is configured with a line length of 120 so I used the same length :)

ramnes commented:
Oh! If flake8 is already configured to 120, then don't mind me.

tych0 commented:
I think I just did that to avoid having to reformat everything. I think we
can definitely have a discussion about shrinking it.

On Thu, Apr 16, 2020, 1:23 PM Guillaume Gelin <notifications@github.com>
wrote:

> Oh! If flake8 is already configured to 120, then don't mind me.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/qtile/qtile/pull/1626#issuecomment-614848545>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAF7VV5QZI7MHPQBN6KAFH3RM5LMVANCNFSM4MJ3EOXA>
> .
>

pawamoy commented:
- should I put the isort command inside of the pep8 tox env instead of its own?
- what line length?

tych0 commented:
Putting it in pep8 is fine. For imports, seems like we should be able to get away with 80 characters pretty easily for a line length.

ramnes commented:
You might want to use https://github.com/gforcada/flake8-isort

pawamoy commented:
Alright I'll update this. I should definitely take a look at all those flake8 plugins :smile:

pawamoy commented:
Some things detected by `flake8-isort` are not fixed by running `isort` itself with the same configuration.

```
./libqtile/widget/battery.py:40:1: I001 isort found an import in the wrong position
./libqtile/widget/battery.py:47:1: I003 isort expected 1 blank line in imports, found 0
./libqtile/widget/battery.py:47:1: I003 isort expected 1 blank line in imports, found 0
./libqtile/widget/generic_poll_text.py:21:1: E303 too many blank lines (3)
```

I fixed them manually.

The first three were due to an import kept apart because of a `noqa` comment.

```python
from typing import Dict  # noqa: F401; noqa: F401
from typing import Any, List, NamedTuple, Optional, Tuple
```

Fixed by

```python
from typing import Any, Dict, List, NamedTuple, Optional, Tuple
```

Not sure what the F401 warning is about.

pawamoy commented on libqtile/images.py:
```diff
@@ -8,6 +8,11 @@
 # The above copyright notice and this permission notice shall be included in
 # all copies or substantial portions of the Software.
 
+import io
+import os
+import re
+from collections import OrderedDict, defaultdict, namedtuple
+
```
Hmm, this doesn't seem right. I think we should concatenate both comments section to the imports are not inserted between the two.

pawamoy commented on test/layouts/test_bsp.py:
```diff
@@ -17,12 +17,13 @@
 # OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 # SOFTWARE.
 
+from test.conftest import no_xinerama
+from test.layouts.layout_utils import assert_focus_path, assert_focused
+
```
I think we should mark `test` as third-party in isort's config. Will fix it.

pawamoy commented:
I'm not sure why but `flake8` does not detect unused imports in the tests, for example `import libqtile.bar` and `import libqtile.scratchpad` in `test_scratchpad.py`.

tych0 commented:
This all looks great to me, thank you for doing it!
